### PR TITLE
feat: multi namespace deployment support

### PR DESCRIPTION
**Pull request by deryrahman**
- [x] merge [proton changes](https://github.com/odpf/proton/tree/deploy-multi-namespace) & update `PROTON_COMMIT` hash

### CODE REVIEW

File: api/handler/v1beta1/job.go [hunk 1]
```diff
@@ -13,46 +14,89 @@ import (
...
-		err = sv.jobSvc.Create(respStream.Context(), namespaceSpec, adaptJob)
+		namespaceSpec, err := sv.namespaceService.Get(stream.Context(), req.GetProjectName(), req.GetNamespaceName())
 		if err != nil {
-			return status.Errorf(codes.Internal, "%s: failed to save %s", err.Error(), adaptJob.Name)
+			stream.Send(&pb.DeployJobSpecificationResponse{
+				Success: false,
+				Ack:     true,
+				Message: err.Error(),
+			})
+			return mapToGRPCErr(sv.l, err, "unable to get namespace")
```
Comment (sbchaos) [inline api/handler/v1beta1/job.go:39]: This will close the stream and not handle rest of jobs. Is it intentional ?

File: api/handler/v1beta1/job.go [hunk 2]
```diff
@@ -13,46 +14,89 @@ import (
...
 
-	observers := new(progress.ObserverChain)
-	observers.Join(sv.progressObserver)
-	observers.Join(&jobSyncObserver{
-		stream: respStream,
-		log:    sv.l,
-		mu:     new(sync.Mutex),
-	})
+		var jobsToKeep []models.JobSpec
+		for _, reqJob := range req.GetJobs() {
```
Comment (sbchaos) [inline api/handler/v1beta1/job.go:43]: We are sending a list of jobs on a bi-directional stream, is this intentional ?

File: config/loader.go [hunk 3]
```diff
@@ -35,18 +43,90 @@ func InitOptimus() (*Optimus, error) {
...
+			namespace.Datastore[i].Path = path.Join(currPath, fileInfo.Name(), d.Path)
+		}
+
+		// assigning to namespaces map
+		namespaces[namespace.Name] = &namespace
+	}
+	return nil
+}
+
+func LoadConfig(cfg interface{}, fs afero.Fs, paths ...string) error {
```
Comment (sbchaos) [inline config/loader.go:109]: Is this used outside of this module? Otherwise it should be `loadConfig`.

File: api/handler/v1beta1/job.go (hunk 1 above)
Comment (deryrahman) [reply api/handler/v1beta1/job.go]: it's not what we want. we've already fixed that https://github.com/odpf/optimus/pull/210/commits/0bba6a4d925fbbb62bc1f17f89f15347e40d671b

File: api/handler/v1beta1/job.go (hunk 2 above)
Comment (deryrahman) [reply api/handler/v1beta1/job.go:43]: Yes, as of now, we send list of jobs per one request. We utilize the bidirectional stream for namespaces level instead of jobs

File: api/handler/v1beta1/resource.go [hunk 4]
```diff
@@ -74,36 +77,84 @@ func (sv *RuntimeServiceServer) ReadResource(ctx context.Context, req *pb.ReadRe
...
+			continue
+		}
+		var resourceSpecs []models.ResourceSpec
+		var errAdaptedResource error
+		for _, resourceProto := range request.GetResources() {
+			adapted, err := sv.adapter.FromResourceProto(resourceProto, request.DatastoreName)
+			if err != nil {
+				sv.l.Error(fmt.Sprintf("%s: cannot adapt resource %s", err.Error(), resourceProto.GetName()))
+				errAdaptedResource = errors.New("resource adapt is failed")
+				continue
```
Comment (sbchaos) [inline api/handler/v1beta1/resource.go:114]: This should be using `break` as we are checking the error below for this loop and using the same error for any failure.

Can be done as part of a separate cleanup pr.

File: cmd/commands.go [hunk 5]
```diff
@@ -112,27 +111,23 @@ func New(plainLog log.Logger, jsonLog log.Logger, conf config.Optimus, pluginRep
...
-	if conf.Namespace.Job.Path != "" {
-		jobSpecRepo = local.NewJobSpecRepository(
-			jobSpecFs,
-			local.NewJobSpecAdapter(pluginRepo),
-		)
-	}
-	datastoreSpecsFs := map[string]afero.Fs{}
-	for _, dsConfig := range conf.Namespace.Datastore {
-		datastoreSpecsFs[dsConfig.Type] = afero.NewBasePathFs(afero.NewOsFs(), dsConfig.Path)
+	datastoreSpecFs := make(map[string]map[string]afero.Fs)
```
Comment (sbchaos) [inline cmd/commands.go:114]: :+1:

File: cmd/deploy.go [hunk 6]
```diff
@@ -95,192 +88,312 @@ func postDeploymentRequest(l log.Logger, projectName string, namespaceName strin
...
+				streamErrs = append(streamErrs, errors.New(resp.GetMessage()))
+			}
+			counter++
+			spinner.SetProgress(counter)
+		}
+	}
+	spinner.Stop()
+	if len(streamErrs) > 0 {
+		for _, e := range streamErrs {
+			fmt.Println(e)
```
Comment (sbchaos) [inline cmd/deploy.go:220]: Use logger instead of Println

File: cmd/deploy.go [hunk 7]
```diff
@@ -95,192 +88,312 @@ func postDeploymentRequest(l log.Logger, projectName string, namespaceName strin
...
-		if err != nil {
-			if errors.Is(err, context.DeadlineExceeded) {
-				l.Error(coloredError("Deployment process took too long, timing out"))
-			}
-			return fmt.Errorf("deployement failed: %w", err)
+	}
+	spinner.Stop()
+	if len(streamErrs) > 0 {
+		for _, e := range streamErrs {
+			fmt.Println(e)
```
Comment (sbchaos) [inline cmd/deploy.go:323]: Use logger

File: cmd/deploy.go [hunk 8]
```diff
@@ -95,192 +88,312 @@ func postDeploymentRequest(l log.Logger, projectName string, namespaceName strin
...
+	datastoreRepo models.DatastoreRepo,
+	datastoreSpecFs map[string]map[string]afero.Fs,
+	namespaceNames []string,
+	verbose bool,
+) error {
+	var selectedNamespaceNames []string
+	if len(namespaceNames) > 0 {
+		selectedNamespaceNames = namespaceNames
+	} else {
+		for _, namespace := range conf.Namespaces {
```
Comment (sbchaos) [inline cmd/deploy.go:239]: Can use spread for values of append, eg
```golang
selectedNamespaceNames = append(selectedNamespaceNames, ...conf.Namespaces)
```

Can be part of separate pr.

File: cmd/deploy.go [hunk 9]
```diff
@@ -95,192 +88,312 @@ func postDeploymentRequest(l log.Logger, projectName string, namespaceName strin
...
-			}
+	ch := make(chan error, len(selectedNamespaceNames))
+	defer close(ch)
+	for _, namespaceName := range selectedNamespaceNames {
+		go func(name string) {
+			ch <- registerNamespace(deployTimeoutCtx, runtime, l, conf, name)
+		}(namespaceName)
+	}
+	var errMsg string
+	for i := 0; i < len(selectedNamespaceNames); i++ {
```
Comment (sbchaos) [inline cmd/deploy.go:351]: Can range over the channel: https://gobyexample.com/range-over-channels

File: config/loader_test.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+	"github.com/odpf/optimus/config"
+
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	configFileName       = ".optimus.yaml"
+	optimusConfigDirName = "./optimus"
+	optimusConfigContent = `
```
Comment (sbchaos) [inline config/loader_test.go:17]: nit: Create a separate const for this one alone.

File: config/loader_test.go [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    path: ./jobs-a
+- name: namespace-b
+  job:
+    path: ./jobs-b
+`
+)
+
+func setup(content string) {
+	teardown()
+	if err := os.Mkdir(optimusConfigDirName, os.ModePerm); err != nil {
```
Comment (sbchaos) [inline config/loader_test.go:58]: `ModePerm` is `0o777`, not good for file permissions. Should use more restrictive permissions.

address in next pr.

File: mock/job.go [hunk 12]
```diff
@@ -128,7 +128,7 @@ type JobService struct {
 }
 
 func (srv *JobService) Create(ctx context.Context, spec2 models.NamespaceSpec, spec models.JobSpec) error {
-	args := srv.Called(ctx, spec, spec2)
+	args := srv.Called(ctx, spec2, spec)
```
Comment (sbchaos) [inline mock/job.go:131]: Better names, spec2 and spec, do not convey any meaning and can easily be wrong.

File: mock/proto_adapter.go [hunk 13]
```diff
@@ -0,0 +1,232 @@
+package mock
+
+import (
+	models "github.com/odpf/optimus/models"
```
Comment (sbchaos) [inline mock/proto_adapter.go:4]: nit: Alias is not required.

File: mock/runtime.go [hunk 14]
```diff
@@ -1,43 +1,248 @@
 package mock
 
 import (
-	"context"
+	context "context"
```
Comment (sbchaos) [inline mock/runtime.go:4]: nit: Alias are not adding more readability

**Review by sbchaos — APPROVED** [review]

File: cmd/deploy.go (hunk 8 above)
Comment (irainia) [reply cmd/deploy.go:239]: What we need is the slice of namespace names, not the namespaces themselves. That's why we iterate it using a loop.

File: cmd/deploy.go (hunk 9 above)
Comment (irainia) [reply cmd/deploy.go:351]: if we range from the channel, at some point we will hit a deadlock error caused by the goroutine not sending anything to the channel, that's why we use this instead

File: cmd/deploy.go (hunk 7 above)
Comment (irainia) [reply cmd/deploy.go]: updated

File: api/handler/v1beta1/resource.go (hunk 4 above)
Comment (irainia) [reply api/handler/v1beta1/resource.go:114]: updated the implementation, though still using the `continue` instead of break because we want to accumulate the error messages.

File: config/loader_test.go (hunk 10 above)
Comment (irainia) [reply config/loader_test.go]: updated with separating it from the other two

File: config/loader_test.go (hunk 11 above)
Comment (irainia) [reply config/loader_test.go]: updated to put more restriction on the permission

File: mock/job.go (hunk 12 above)
Comment (irainia) [reply mock/job.go]: updated the naming for better expressing intent

File: mock/runtime.go (hunk 14 above)
Comment (irainia) [reply mock/runtime.go]: updated by removing it

File: mock/proto_adapter.go (hunk 13 above)
Comment (irainia) [reply mock/proto_adapter.go]: updated by removing it
